interop-tests/src/test/scala/akka/grpc/scaladsl/PowerApiSpec.scala

```diff
@@ -84,5 +117,49 @@ class PowerApiSpec(backend: String)
       response.headers.getText("baz").get should be("qux")
       response.trailers.futureValue.getText("foo").get should be("bar")
     }
+
+    "(on streamed calls) redeem the headers future as soon as they're available (and trailers future when trailers arrive)" in {
+
+      // invoking streamed calls using the power API materializes a Future[GrpcResponseMetadata]
+      // that should redeem as soon as the HEADERS is consumed. Then, the GrpcResponseMetadata instance
+      // contains another Future that will redeem when receiving the trailers.
+      client = GreeterServiceClient(
+        GrpcClientSettings.connectToServiceAt("localhost", server.localAddress.getPort).withTls(false))
+
+      val p = Promise[HelloRequest]
+      val requests: Source[HelloRequest, NotUsed] = Source.single(HelloRequest("Alice")).concat(Source.future(p.future))
+
+      val responseSource: Source[HelloReply, Future[GrpcResponseMetadata]] =
+        client.streamHellos().invokeWithMetadata(requests)
+
+      val minNumberOfResponses = 200
+      val latch = new CountDownLatch(minNumberOfResponses)
+      val i: AtomicInteger = new AtomicInteger(0)
+      val headers: Future[GrpcResponseMetadata] = responseSource
+        .map { x =>
+          i.incrementAndGet()
+          latch.countDown()
+          x
+        }
+        .to(Sink.ignore)
+        .run
+
+      // blocks progress until redeeming `headers`
+      val trailers = headers.futureValue.trailers
+
+      // Don't send the finalization message until the headers future was redeemed (see above)
+      // and we've received a good amount of responses.
+      latch.await(5, TimeUnit.SECONDS)
+      trailers.isCompleted should be(false)
+      p.trySuccess(HelloRequest("ByeBye"))
```

Comment:
As suggested by @raboof, the server doesn't finish until a marking message (in this case `HelloRequest("ByeBye")`) is received, and that finalization message is not sent until `headers` is redeemed.